### PR TITLE
Fix invalid IR being created by CheriBoundAllocas pass

### PR DESCRIPTION
**Pull request by arichardson**
When we encounter multiple uses within the same instruction, we need
to ensure that we reuse the same bounded alloca intrinsic. If we
don't do this, we could end up creating invalid PHI nodes where the
PHI node has multiple entries for the same basic block and uses
different incoming values (for an example see
multiple-uses-in-same-instr.ll). This not only avoids invalid IR, but
also avoids unnecessarily creating multiple intrinsic calls e.g.
in cases where a call instruction passes the same IR variable twice.

### CODE REVIEW

File: llvm/test/Transforms/CheriBoundAllocas/broken-phi.ll [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  switch i32 %arg, label %cond.end.sink.split [
+  i32 -1, label %cond.end
+  i32 0, label %cond.end
+  ]
+
+cond.end.sink.split:
+  br label %cond.end
+
+cond.end:
+  %ref.tmp109.sink = phi %class.QString addrspace(200)* [ %ref.tmp107, %cond.true ], [ %ref.tmp107, %cond.true ], [ null, %cond.end.sink.split ]
```
Comment (jrtc27) [inline llvm/test/Transforms/CheriBoundAllocas/broken-phi.ll]: I'm surprised this is legal IR to be honest, though maybe the lesser of two evils

File: llvm/test/Transforms/CheriBoundAllocas/multiple-uses-in-same-instr.ll [hunk 2]
```diff
@@ -13,11 +13,11 @@ target triple = "riscv64-unknown-freebsd13"
 %struct.QTypedArrayData = type { %struct.QArrayData }
```
Comment (jrtc27) [inline llvm/test/Transforms/CheriBoundAllocas/multiple-uses-in-same-instr.ll:10]: Fold this commit into the original test case one?

File: llvm/test/Transforms/CheriBoundAllocas/broken-phi.ll (hunk 1 above)
Comment (arichardson) [reply llvm/test/Transforms/CheriBoundAllocas/broken-phi.ll]: Yeah I'm not quite sure what the original was before llvm-reduce. Maybe it looked a bit more sane.

File: llvm/test/Transforms/CheriBoundAllocas/multiple-uses-in-same-instr.ll (hunk 2 above)
Comment (arichardson) [reply llvm/test/Transforms/CheriBoundAllocas/multiple-uses-in-same-instr.ll:10]: Any further suggestions for this PR? If not I'll address this comment and merge then.

File: llvm/include/llvm/Analysis/CheriBounds.h [hunk 3]
```diff
@@ -33,7 +34,7 @@ class CheriNeedBoundsChecker {
                                  const APInt &CurrentGEPOffset,
                                  unsigned Depth) const;
 
-  const Instruction *I;
+  const Instruction *RootInst;
```
Comment (jrtc27) [inline llvm/include/llvm/Analysis/CheriBounds.h:37]: We should probably stop making this const so UsesThatNeedBounds can contain non-const uses and avoid needing to const_cast when updating them, but that's a separate issue...

File: llvm/lib/CodeGen/CheriBoundAllocas.cpp [hunk 4]
```diff
@@ -295,44 +295,60 @@ class CheriBoundAllocas : public ModulePass, public InstVisitor<CheriBoundAlloca
...
-        } else {
-          if (auto PHI = dyn_cast<PHINode>(I)) {
-            // For PHI nodes we can't insert just before the PHI, instead we
-            // must insert it just before
-            auto BB = PHI->getIncomingBlock(*U);
-            // LLVM_DEBUG(dbgs() << "PHI use coming from"; BB->dump());
-            B.SetInsertPoint(BB->getTerminator());
+        }
+      } else {
+        // Otherwise, we create new intrinsics for every use. This can avoid
```
Comment (jrtc27) [inline llvm/lib/CodeGen/CheriBoundAllocas.cpp:311]: Hm, why do we actually bother doing this? If it's rematerialisable, shouldn't it not matter?

File: llvm/lib/CodeGen/CheriBoundAllocas.cpp [hunk 5]
```diff
@@ -295,44 +295,60 @@ class CheriBoundAllocas : public ModulePass, public InstVisitor<CheriBoundAlloca
...
+              auto BB = PHI->getIncomingBlock(*U);
+              // LLVM_DEBUG(dbgs() << "PHI use coming from"; BB->dump());
+              B.SetInsertPoint(BB->getTerminator());
+            } else {
+              // Insert just before the use. This should avoid spilling
+              // registers when using an alloca in a different basic block.
+              B.SetInsertPoint(I);
+            }
+            // We need to convert it to an i8* for the intrinisic. Note: we have
+            // to create a new bitcast every time since reusing the same one can
```
Comment (jrtc27) [inline llvm/lib/CodeGen/CheriBoundAllocas.cpp:340]: This seems... odd; bitcasts _should_ be no-ops that get lost way before register allocation...

File: llvm/lib/CodeGen/CheriBoundAllocas.cpp (hunk 4 above)
Comment (arichardson) [reply llvm/lib/CodeGen/CheriBoundAllocas.cpp:311]: Good point. I'm afraid I don't remember exactly why it needs to be done here. Doing it later should be better but there were some issues with rematerialization, so I ended up manually doing it in this pass.

Comment (arichardson) [reply llvm/lib/CodeGen/CheriBoundAllocas.cpp:311]: Hmmm maybe that logic predates `CheriBoundedStackPseudo`, or maybe the MIR rematerialization caused some bad codegen...

File: llvm/test/CodeGen/CHERI-Generic/Inputs/stack-bounds-remat.ll [hunk 6]
```diff
@@ -0,0 +1,40 @@
+; Normally llc would explicitly insert stack bounds, but we want to skip this pass here
```
[the remark this answers is not in the document]
Comment (jrtc27) [reply llvm/test/CodeGen/CHERI-Generic/Inputs/stack-bounds-remat.ll:1]: Yeah, I don't think LLVM takes into account the fact that it has to spill a callee-saved register in its register allocation, only considers rematerialisability when it has to spill the rematerialisable value itself

**Comment by arichardson** [comment]
Dropped the test, hopefully fine to merge now?

**Pull request merged**